File: src/subcommands/subcommands.cpp
#include "subcommands.h"
#include "main.h"
#include "start.h"
#include "fill.h"
#include "log.h"
#include "remove.h"
#include "summary.h"

namespace {
    std::vector<subcommands::subcommand*>* vector = nullptr;
}

std::vector<subcommands::subcommand*>& subcommands::subcommands()
{
    if (vector == nullptr) {
        vector = new std::vector<subcommands::subcommand*>{
            new subcommands::main(),
            new subcommands::start(),
            new subcommands::fill(),
            new subcommands::log(),
            new subcommands::remove(),
            new subcommands::summary()
        };
    }

    return *vector;
}

subcommands::subcommand* subcommands::find(const std::string& command)
{
    for (subcommands::subcommand* subcommand : subcommands::subcommands()) {
        if (subcommand->command() == command) {
            return subcommand;
        }
    }

    return nullptr;
}

File: src/subcommands/subcommands.rs
use std::error::Error;
use clap::{Clap, crate_version, crate_description};
use enum_dispatch::enum_dispatch;
use super::{Fill, Log, Remove, Start, Summary, Count, GenerateCompletions};

#[enum_dispatch]
pub trait Subcommand {
    /// Runs the subcommand with the parsed arguments.
    fn run(&self) -> Result<(), Box<dyn Error>>;
}

#[enum_dispatch(Subcommand)]
#[derive(Clap)]
#[clap(version = crate_version!(), about = crate_description!())]
pub enum Subcommands {
    Log(Log),
    Start(Start),
    Fill(Fill),
    Remove(Remove),
    Count(Count),
    Summary(Summary),
    GenerateCompletions(GenerateCompletions)
}
